### PR TITLE
Work around potential race in PipeWriter Redux

### PR DESCRIPTION
**Pull request by halter73**
Maybe we'll address #8843 this time.

### CODE REVIEW

**Comment by jkotalik** [comment]
Can you try to rerun `AngularTemplate_Works` multiple times as that test repro'd the race fairly consistently?

**Comment by jkotalik** [comment]
> /home/vsts/work/1/s/src/Servers/Kestrel/perf/Kestrel.Performance/Http1LargeWritingBenchmark.cs(14,53): error CS0234: The type or namespace name 'Abstractions' does not exist in the namespace 'Microsoft.AspNetCore.Server.Kestrel.Transport' (are you missing an assembly reference?) [/home/vsts/work/1/s/src/Servers/Kestrel/perf/Kestrel.Performance/Microsoft.AspNetCore.Server.Kestrel.Performance.csproj]

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimingPipeFlusher.cs [hunk 1]
```diff
@@ -22,7 +22,13 @@ internal class TimingPipeFlusher
         private readonly IKestrelTrace _log;
 
         private readonly object _flushLock = new object();
-        private Task<FlushResult> _lastFlushTask = null;
+
+        // This ValueTask that was either:
+        // 1. The default value which "IsCompleted"
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimingPipeFlusher.cs]:
```suggestion
        // 1. The default value where "IsCompleted" is true
```

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimingPipeFlusher.cs]: Also, comment on how _lastFlushTask must only be get/set under the _flushLock.

**Review by jkotalik — APPROVED** [review]
Really make sure `AngularTemplate_Works` passes multiple times.

**Review by ryanbrandenburg — COMMENTED** [review]
Templates parts seem fine. Someone with Kestrel knowledge should give final approval.

File: src/ProjectTemplates/test/SpaTemplateTest/AngularTemplateTest.cs [hunk 2]
```diff
@@ -17,8 +17,13 @@ public AngularTemplateTest(ProjectFactoryFixture projectFactory, BrowserFixture
             : base(projectFactory, browserFixture, output) { }
 
         [Fact]
-        public Task AngularTemplate_Works()
```
Comment (jkotalik) [inline src/ProjectTemplates/test/SpaTemplateTest/AngularTemplateTest.cs]: Next time, just add [Repeat(10)] 😉 .

As long as you revert this after it all passes, 🚢 🇮🇹

**Comment by jkotalik** [comment]
I got a feeling the template tests are not fond of being run multiple times in a row.

File: src/Shared/Buffers.MemoryPool/MemoryPoolFactory.cs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 namespace System.Buffers
 {
-    internal static class MemoryPoolFactory
+    internal static class SlabMemoryPoolFactory
```
Comment (davidfowl) [inline src/Shared/Buffers.MemoryPool/MemoryPoolFactory.cs]: LOL

Comment (halter73) [reply src/Shared/Buffers.MemoryPool/MemoryPoolFactory.cs]: You can't say I didn't warn you :trollface:

**Review by davidfowl — APPROVED** [review]
What's the performance delta?

**Comment by halter73** [comment]
> What's the performance delta?

I can see no reason for it to be any different than https://github.com/aspnet/AspNetCore/pull/10165#issuecomment-491469419, and that took a while to run the first time.

**Pull request merged**